### ISSUE TITLE
System.Console unexpectedly uses a UTF-8 encoding *with BOM* on Unix

### ISSUE THREAD

**Issue by mklement0** [dotnet/corefx#32004]
On Unix-like platforms, the character encoding to use by default in `System.Console` instances is determined as follows:

* If any of the following environment variables are set (which is the norm), the character encoding is derived from the first one that is defined: `LC_ALL`, `LC_MESSAGES`, `LANG`); e.g., if
`LANG="en_US.UTF-8"` is present, string `"UTF-8"` is extracted and the `Encoding` instance is obtained by [passing that string to `Encoding.GetEncoding()`](https://github.com/dotnet/corefx/blob/a10890f4ffe0fadf090c922578ba0e606ebdd16c/src/Common/src/System/Text/EncodingHelper.Unix.cs#L16).

* Otherwise, the default is _BOM-less_ UTF-8, as expected.

The problem is that calling `Encoding.GetEncoding("UTF-8")` yields a UTF-8 encoding _with BOM_, whereas Unix-like platforms neither expect nor do most utilities there know how to process a UTF-8 BOM.

Here's a quick demonstration in PowerShell:

```powershell
PS> [Text.Encoding]::GetEncoding('UTF-8').GetPreamble().Count
3  # encoding *with BOM* (3 bytes) was returned.
```

As an aside: Curiously, even though on _Windows_ the BOM is also present, switching to UTF-8 in a `cmd.exe` console with `chcp 65001` and using `>` to redirect a non-ASCII string to a file results in a UTF-8-encoded file _without_ BOM.

**Comment by tmds** [comment]
> Curiously, even though on Windows the BOM is also present, switching to UTF-8 in a cmd.exe console with chcp 65001 and using > to redirect a non-ASCII string to a file results in a UTF-8-encoded file without BOM.

Redirecting on Unix also gives a file without a BOM. The Console strips the BOM from the encoding for output:

https://github.com/dotnet/corefx/blob/775c9f1415be3403864fe2804a9c5e16074d2d68/src/System.Console/src/System/Console.cs#L116-L125

We can do the same thing to the Console.InputEncoding and Console.OutputEncoding.

**Comment by karelz** [comment]
Fixing milestone - this was fixed in master (3.0) branch.

**Comment by mklement0** [comment]
In hindsight I shouldn't have made this about Unix only - I think  that **on Windows too the preamble (BOM) should be removed**, given that `cmd.exe` - rightfully - has always operated _without BOM_ with code page `65001` in effect:

After all, **any programs relying on the _active code page_ should be able to _blindly_ assume that stdin input / stdout output  is encoded accordingly and _shouldn't have to deal with a BOM_** (neither to redundantly imply the same code page nor to signal a different encoding).

The presence of a BOM, in fact, breaks Windows PowerShell's background-jobs feature (via `Start-Job`, as detailed here:

* https://windowsserver.uservoice.com/forums/301869-powershell/suggestions/14915283-job-cmdlets-fail-with-utf-8-codepage

Follow-on bugs:

* https://github.com/Microsoft/azure-pipelines-tasks/issues/9766
* https://github.com/ansible/ansible/issues/15770

---

Demonstration of `cmd`'s current behavior:

`cmd`'s stdout output has always created _BOM-less_ output with `chcp 65001` in effect, as stated in the OP; e.g. the following creates `t.txt` as a _BOM-less_ UTF-8 file.

```powershell
C:\> chcp 65001
C:\> (echo hü)> t.txt
C:\> powershell -noprofile -c Format-Hex t.txt

           Path: C:\Users\jdoe\t.txt

           00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F

00000000   68 C3 BC 0D 0A                                   hÃ¼..

```

The above shows that the file contains _no_ BOM (and that [char. `ü`](http://www.fileformat.info/info/unicode/char/fc/index.htm) was correctly encoded as 2-byte sequence `0xC3 0xBC`).

**Comment by mklement0** [comment]
I've since realized that this should be fixed on Windows as well - please see #35950.
